File: fs/Consumer.java
package fs;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Consumer {

	public static void main(String[] args) {
		BufferedReader in = null;
		int c;
		char read;
		
		if (args.length < 1 || args.length > 2) { //Invocation error
			System.out.println("usage: java Consumer <params> [filename]");
			System.out.println("OR");
			System.out.println("usage: java Consumer <params> < [filename]");
			System.exit(0);
		}
		else if (args.length == 1) { // only params
			in = new BufferedReader(new InputStreamReader(System.in));		
		}
		else if (args.length == 2) {// params and filename
			try {
				in = new BufferedReader(new FileReader(args[1]));
				System.out.println("Reading file...");
			}catch(FileNotFoundException e){
				System.out.println("File not found");
				System.exit(1);
			}
		}

		try {
			// Filter by character
			while ((c = in.read()) > 0) {
				read = (char)c;
				
				// write here your logic

				// end logic

				System.out.print(read);
			}
		} catch(IOException ex){
			System.out.println("Errore di input");
			System.exit(2);
		}
	}
		
}

File: fs/consumer.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MAX_STRING_LENGTH 256

int main(int argc, char *argv[]) {
	char *file_in, *prefix;
	char err[MAX_STRING_LENGTH], buffer[MAX_STRING_LENGTH];

	int nread, i, fd;
	char read_char;

	// CHECK ARGS
	if (argc < 2 || argc > 3) {
		printf("error: %s prefix < filename \n OR \n %s prefix filename\n", argv[0], argv[0]);
		return EXIT_FAILURE;
	}

	else if (argc == 2) {
		printf("invocation without filename passed...\n");
		prefix = argv[1];

		// 0 Standard input (stdin)
		fd = 0;
	}
	else if (argc == 3) {
		printf("invocation with filename passed...\n");
		prefix = argv[1];
		file_in = argv[2];

		fd = open(file_in, O_RDONLY);
		if (fd < 0) {
			perror("p0: cannot open selected file");
			return EXIT_FAILURE;
		}
	}

	while ((nread = read(fd, &read_char, sizeof(char)))) { // till EOF
		if (nread < 0) {
			perror("error: somthing bad happened while reading file");
			return EXIT_FAILURE;
		}

		// write here your logic

		// end logic

		printf("%c", read_char);

	}

	close(fd);
	return EXIT_SUCCESS;
}
